dnssec-backend.go

```diff
@@ -0,0 +1,313 @@
+		case *dns.RRSIG:
+			result.RrSig = t
+		case *dns.DNSKEY, *dns.DS, *dns.A, *dns.AAAA:
+			result.RrSet = append(result.RrSet, rr)
+		}
+	}
+	return result, nil
+}
+
+func doQuery(q *dns.Msg, resolver Resolver, ci ClientInfo) (*dns.Msg, error) {
+	r, err := resolver.Resolve(q, ci)
+	if err != nil {
+		return nil, err
+	}
+	if r.Rcode == dns.RcodeNameError {
+		Log.Info("no such domain", "info", qName(r))
+		return nil, ErrNoResult
+	}
+	if r == nil || r.Rcode == dns.RcodeSuccess {
+		return r, err
+	}
+	return nil, ErrInvalidRRsig
+}
+
+func queryDelegation(domainName string, resolver Resolver, ci ClientInfo) (*SignedZone, error) {
+	signedZone := &SignedZone{
+		Zone:   domainName,
+		Ds:     &RRSet{},
+		Dnskey: &RRSet{},
+	}
+	signedZone.PubKeyLookup = make(map[uint16]*dns.DNSKEY)
+
+	var g errgroup.Group
+	g.Go(func() error {
+		var err error
+		signedZone.Dnskey, err = getRRset(domainName, dns.TypeDNSKEY, resolver, ci)
+		if err != nil {
+			return err
+		}
+
+		for _, rr := range signedZone.Dnskey.RrSet {
+			defer func() {
+				if r := recover(); r != nil {
+					Log.Warn("panic occurred",
+						slog.String("domainName", domainName),
+						slog.Any("panic", r),
+					)
+				}
+			}()
+			signedZone.addPubKey(rr.(*dns.DNSKEY))
+		}
+		return nil
+	})
+
+	g.Go(func() error {
+		var err error
+		signedZone.Ds, err = getRRset(domainName, dns.TypeDS, resolver, ci)
+		if err != nil {
+			Log.Error("DS query failed",
+				slog.String("domainName", domainName),
+				"error", err,
+			)
+
+			return err
+		}
+		return nil
+	})
+
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+	return signedZone, nil
+}
+
+func (authChain *AuthenticationChain) Populate(domainName string, resolver Resolver, ci ClientInfo) error {
+	qnameComponents := strings.Split(domainName, ".")
+	zonesToVerify := len(qnameComponents)
+	authChain.DelegationChain = make([]SignedZone, zonesToVerify)
+
+	var g errgroup.Group
+	for i := 0; i < zonesToVerify; i++ {
+		zoneName := dns.Fqdn(strings.Join(qnameComponents[i:], "."))
+		index := i
+
+		g.Go(func() error {
+			delegation, err := queryDelegation(zoneName, resolver, ci)
+			if err != nil {
+				return err
+			}
+			authChain.DelegationChain[index] = *delegation
+			if index > 0 {
+				authChain.DelegationChain[index-1].ParentZone = delegation
+			}
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (authChain *AuthenticationChain) Verify(answerRRset *RRSet) error {
+	zones := authChain.DelegationChain
+	if len(zones) == 0 {
+		return ErrDelegationChain
+	}
+
+	signedZone := authChain.DelegationChain[0]
+	if !signedZone.checkHasDnskeys() {
+		return ErrDnskeyNotAvailable
+	}
+
+	err := signedZone.verifyRRSIG(answerRRset)
+	if err != nil {
+		return ErrInvalidRRsig
+	}
+
+	for _, signedZone := range authChain.DelegationChain {
+		defer func() {
+			if err := recover(); err != nil {
+				Log.Error("AuthChain panic occurred", "error", err)
+			}
+		}()
```

Comment:
Again, there should be a better way to deal with panics, i.e. avoid them at all.